src/libraries/System.Private.Uri/src/System/Uri.cs

```diff
@@ -2377,11 +2378,12 @@ private unsafe void CreateUriInfo(Flags cF)
         Done:
             cF |= Flags.MinimalUriInfoSet;
             info.DnsSafeHost = _dnsSafeHost;
-            lock (_string)
+
+            Interlocked.CompareExchange(ref _info, info, null!);
```

Comment:
What does `null!` mean? a non-null null? If compare exchange required the comparand not to be null i'd say it should be relaxed since it's entirely legitimate to use null here.

Reply:
The problem here is that `_info` is marked non-nullable (even though it can be null until `CreateUriInfo` is called), so the T in `ComapreExchange` is non-null as well.

Reply:
Right, the problem is we're already lying in one place related to this, so we basically have to lie again.

That said, CompareExchange is already special-cased by the C# compiler for nullability, because it's very commonly used and there are things about it we can't express in the nullable annotations or attributes.  Ideally we could write this instead as:
```C#
Interlocked.CompareExchange<object?>(ref _info, info, null);
```
and it shouldn't warn, as the compiler could see that `info` here wouldn't be null, and thus this wouldn't be trying to assign null to a non-nullable location.  However, we do get a warning, e.g.
```C#
#nullable enable
using System.Threading;

class Program
{

    private static object _obj = null!;
    static void Main() => Interlocked.CompareExchange<object?>(ref _obj, new object(), null);
}
```
warns:
```
warning CS8601: Possible null reference assignment.
```
@jcouv, any thoughts on further special-casing CompareExchange so that regardless of whether the first argument is a ref to a nullable or non-nullable location, the compiler won't warn about a possible null assignment to it if the second argument is seen to be not-null?

Reply:
It's the third argument that concerned me really, It shouldnt need the `!`  It doesn't look like it needs to be specified as non-nullable so can it be removed? Using a non-nullable null is nonsensical.

Reply:
> t's the third argument that concerned me really

I understand that.  It needs the `!` because the type of the generic T is being inferred as non-nullable; hence `null` will get a warning.  If you instead specify `T` as being nullable, you hit the warning I refer to above.

> It doesn't look like it needs to be specified

It does.

> can it be removed?

No

Reply:
We can choose between
```c#
Interlocked.CompareExchange(ref _info, info, null!);
```
and
```c#
Interlocked.CompareExchange<UriInfo?>(ref _info!, info, null);
```

As _info is `UriInfo` (non-nullable).

If _info was `UriInfo?` then
```c#
Interlocked.CompareExchange(ref _info, info, null);
```

would be fine.

Reply:
Right.

Reply:
> any thoughts on further special-casing CompareExchange so that regardless of whether the first argument is a ref to a nullable or non-nullable location, the compiler won't warn about a possible null assignment to it if the second argument is seen to be not-null?

We're considering some more special-casing (particularily for `Task<T>` and task-like types, and `.Where`). Feel free to file a csharplang issue.